iot-edge/EdgeSolution/modules/CSharpModule/Program.cs

```diff
@@ -0,0 +1,172 @@
+            Console.CancelKeyPress += (sender, cpe) => cts.Cancel();
+            WhenCancelled(cts.Token).Wait();
+        }
+
+        /// <summary>
+        /// Handles cleanup operations when app is cancelled or unloads
+        /// </summary>
+        public static Task WhenCancelled(CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSource<bool>();
+            cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).SetResult(true), tcs);
+            return tcs.Task;
+        }
+
+        /// <summary>
+        /// Initializes the ModuleClient and sets up the callback to receive
+        /// messages containing temperature information
+        /// </summary>
+        static async Task Init()
+        {
+            AmqpTransportSettings amqpSetting = new AmqpTransportSettings(TransportType.Amqp_Tcp_Only);
+            ITransportSettings[] settings = { amqpSetting };
+
+            // Open a connection to the Edge runtime
+            ModuleClient ioTHubModuleClient = await ModuleClient.CreateFromEnvironmentAsync(settings);
+            await ioTHubModuleClient.OpenAsync();
+            Console.WriteLine("IoT Hub module client initialized.");
+
+            // Register a callback for messages that are received by the module.
+            // await ioTHubModuleClient.SetImputMessageHandlerAsync("input1", PipeMessage, iotHubModuleClient);
+
+            // Read the TemperatureThreshold value from the module twin's desired properties
+            var moduleTwin = await ioTHubModuleClient.GetTwinAsync();
+            var moduleTwinCollection = moduleTwin.Properties.Desired;
+            try
+            {
+                temperatureThreshold = moduleTwinCollection["TemperatureThreshold"];
+            }
+            catch (ArgumentOutOfRangeException e)
+            {
+                Console.WriteLine($"Property TemperatureThreshold not exist: {e.Message}");
+            }
+
+            // Attach a callback for updates to the module twin's desired properties.
+            await ioTHubModuleClient.SetDesiredPropertyUpdateCallbackAsync(OnDesiredPropertiesUpdate, null);
+
+            // Register a callback for messages that are received by the module.
+            await ioTHubModuleClient.SetInputMessageHandlerAsync("input1", FilterMessages, ioTHubModuleClient);
+        }
+        static Task OnDesiredPropertiesUpdate(TwinCollection desiredProperties, object userContext)
+        {
+            try
+            {
+                Console.WriteLine("Desired property change:");
+                Console.WriteLine(JsonConvert.SerializeObject(desiredProperties));
+
+                if (desiredProperties["TemperatureThreshold"] != null)
+                    temperatureThreshold = desiredProperties["TemperatureThreshold"];
+
+            }
+            catch (AggregateException ex)
+            {
+                foreach (Exception exception in ex.InnerExceptions)
```

Comment:
See https://docs.microsoft.com/en-us/dotnet/api/system.aggregateexception.handle?view=netframework-4.7.2 for the proper way to handle AggregateExceptions.
(Still, I doubt this will ever be thrown from here - if it will, we need to fix that bug.)